doc/gnugol.man
.TH gnugol 1  "January 02, 2011" "version 1" "USER COMMANDS"
.SH NAME
gnugol \- search web based search engines at the command line
.SH SYNOPSIS
.B gnugol
[\-?] [\-o output_format] [\-e engine] [keywords]
.SH DESCRIPTION
The modern 
.B Web 
is a vast, wonderful resource for all kinds of
information. Searching it is a great IQ enhancer.
.PP
The modern 
.B search results 
page, on the other hand, is a vast source of colorized formatting,
innumerable fonts, javascript, multiple images, potential malware, and
advertising, with an ever changing look and feel based on the whim of
whatever graphic designer has a bee up their posterior that month, and
whatever new html features are trendy, with the answer(s) to your
question(s) buried somewhere deep within all that.
.PP
Gnugol is written in the spirit of "whois" and other stable command
line tools in that it always outputs a web search in the format you
want, whether that be html, org, markdown, textile, or plain text.

It outputs a basic combination of "url", "title", and "snippet" in
those formats.
.PP
A web browser is 
.B NOT 
required to use gnugol.
.PP
By outputting results in the format you want, with minimal (no)
markup, gnugol makes it easier to get your answers or create useful
content of your own, in the mode you are used to working in.
.PP
Gnugol reduces shiny web browser distractions while searching.
.PP
There is a gnugol.el for GNU Emacs included with the distribution that
makes many of gnugol's features available in Emacs buffers.
.PP
As it is written in C, gnugol is also lightweight and pretty fast.
.SH OPTIONS
.TP
\-? \-\-help display help text. The features listed via this option are likely to be more up to date than this manual page!
.TP
\-e \-\-engine        [bing|google|dummy]

At some point more engines (citeseer, wikipedia, omega, etc) will be supported. To see the list of supported engines with this build, do a 
.PP
.B gnugol
--about engines
.PP
(The dummy engine is useful for testing)
.TP 
\-o \-\-output        [html|org|mdwn|wiki|text|term|textile]

The default output format is currently org. (The emacs based org-mode outliner format, for Emacs) After utf-8 word breaking is figured out it will probably default to term. More output formats will exist one day.

.TP
\-n \-\-nresults      number of results to fetch

The default is 5.

.TP
\-p \-\-position      start of results to fetch

The default is 0.
.TP
\-U \-\-escaped       input is already url escaped
Gnugol expects keywords in UTF-8 format, shell escaped, by default. A shell escaped AND url escaped argument is often simpler.
.TP
\-l \-\-language-in   [en|es|fr|nl or other 2 char locale]
.TP
\-L \-\-language-out  [en|es|fr|nl or other 2 char locale]
.TP
\-H \-\-header    0|1 
disable/enable output header
.TP
\-F \-\-footer    0|1 
disable/enable output footer
.TP
\-s \-\-snippets 0|1
disable/enable snippet output. Default: enabled.
.TP
\-t \-\-titles 0|1
disable/enable title output. Default: enabled.
.TP
\-u \-\-urls 0|1
disable/enable url output. Default: enabled.
.TP
-i --indent 0..5 
Amount of indentation of results
.TP
\-S \-\-safe 0|1|2 
Set the pr0n-safety of the search. 
.PP
0 = Depraved. 2 = Prudish.
.PP
Default: 1.
.TP
\-v \-\-verbose       provide more verbose insight
.TP
\-d \-\-debug X
Debug level. 1 = only catastrophic errors 11 = everything
.TP
\-a \-\-about [credits|copyright|license|source|config|keys|all]

.SH EXAMPLES
.TP
Basic search:
.B gnugol
in the beginning was the command line
.PP
.TP
Complex search, using google's site restriction parameter:
.B gnugol
\-e google -o mdwn site:nex-6.taht.net latency
.TP
Search bing, output results to elinks
.B gnugol
\-e bing -o html -H 1 -F 1 the Täht bug | elinks
.TP
For more examples see the web site at http://gnugol.taht.net
.PP
.SH EXIT STATUS
gnugol returns a negative exit status if the query fails.
.SH NOTES
To search bing, a license key is required. To search google, no
license key is required, however, it is highly recommended.

A google search is thus, currently, the default.
For more details:

.B gnugol
\-\-about keys

.SH BUGS
Lots! UTF\-8 support is incomplete at the moment. SSML is untested.
.SH AUTHOR
Michael ("Dave") Täht (d+gnugol (at) taht.net) 
.SH SEE ALSO
surfraw(1) whois(1) dig(1) nslookup(1) elinks(1) host(1) 


doc/history.org
#+TITLE:     GNUGOL's history
#+AUTHOR:    Dave Täht
#+EMAIL:     d AT taht.net
#+DATE:      <2010-12-21 Tue>
#+LANGUAGE:  en
#+TEXT:      Searching the Web in Plain Text
#+OPTIONS:   H:3 num:t toc:t \n:nil @:t ::t |:t ^:t -:t f:t *:t TeX:t LaTeX:nil skip:nil d:t tags:not-in-toc
#+INFOJS_OPT: view:nil toc:t ltoc:t mouse:underline buttons:0 path:org-info.js
#+LINK_UP: index.html
#+LINK_HOME: index.html
#+STYLE:    <link rel="icon" type="image/ico" href="http://gnugol.taht.net/images/favicon.ico">
#+STYLE:    <link rel="stylesheet" type="text/css" href="worg.css" />
* Where did the idea come from?
  Dave has had a long experience with search engines, having written one in 1996, and another in 1999. 
** Genesis 
   Gnugol began life in 1996 as a shell script that scraped altavista's search results into a format that was compatible with the links text-mode web browser. It went through various incarnations over the years.
** Frustration with "truth in advertising"
   In a meeting with some web advertising executives in 1999, Dave “lost it”, and started [[http://the-edge.blogspot.com/2003_09_14_archive.html][describing the new IPv13 protocol]], which would guarantee that after so many data packets, “ADV” (advertising) packets had to be sent. 

Everybody in the room bought it, to his dismay, even a few that shouldn't have been fooled. 

Soon afterwards he started working on [[http://www.privoxy.org/][privoxy]] and other tools like that. 
** A trip to Belize
   where he noticed how slow the internet was in remote locations. 
** ISC.org
   Some discussions about the efficiency of dns and udp over ipv6. 
** Trips to Nicaragua and Australia in 2008
   Where it was painfully obvious that the Internet, especially over lossy 802.11 networks, was often too slow to actually use. (Now, it turns out that a key problem is actually [[http://gettys.wordpress.com/2011/01/03/aggregate-bufferbloat-802-11-and-3g-networks/][bufferbloat]])
** Latency
(Note this is just some random math as I try to get mathml to work for me)

 \(l * 3\)

* Early life (1996-2006)
A zillion different primitive shell scripts, perl based scrapers, meta search engines, etc. To search the most popular search engines at the time, notably, using text based browsing tools like links, lynx, and elinks.
* Adolescence (march, 2008)
** Combating Featuritus
In the race to add feature after feature after freature, google and other search engines have lost sight of the original appeal of their services - getting an answer to a question. 

A typical google page is now over 25K long, not counting the image, contains dozens of extra links, and takes 5-10 seconds to download and render over slow, international links. Not only that but the various search engines are contributing to breaking the Internet, by, for example, increasing the default MSS size to 10 from 2...

Dave had always admired the DNS system, and felt that now that "search" is almost a commodity, that it would be possible to define a binary (udp) based protocol for search, this time using a flexible header format, and compression and encryption of that. There are plenty of unused, useless services in /etc/services - having a special port number for search makes sense.

Problem with that was that most query-response protocols don't work well through a firewall. 

Enter IPv6. Without NAT, you have end to end networking.  Response problem solved. Billing problem solved, too. Or so he thought. 

** Gaining speed
The original version of the gnugold network protocol used 2!! packets to get a query out and a response. If you went through a 6in4 tunnel, it was 4 packets. (2 very short ones, however)

This means that by the time you completed a tcp handshake vs the normal google, you had already got a response from gnugold, on a long latency link. 

The earliest prototype came in at *half* the time to transmit a query from australia and get the response that google.com.au did, best case. Due to the unreliable network I was on, it was actually often 5-10 times faster. 

Additional benefits - vastly reduced data traffic. Control of the formatting can be controlled on the phone or remote servers. 

38 packets vs 4. Not bad. Some additional tweaking was in order.

The second nice thing about end to end networking is that it makes it easier -assuming a static ip address - to have for-pay search. Dave would gladly pay a few bucks a month for faster - and ad-free searching!

* Death #1
Unfortunately the UDP based network system described above proved unreliable enough in practice for Dave to discard it (at least temporarily) in favor of a few other network backend ideas. (See also: [[http://gettys.wordpress.com/category/bufferbloat/][Bufferbloat]] ) The ad-hoc protocol was also a terrible design, too close to HTTP in origination.
* Death #2
And THEN google obsoleted their soap API in favor of json, rendering the original front-end client useless.
* Death #3
Finally that soap API went away...
* Rebirth - October, 2010
After encountering [[http://nex-6.taht.net/posts/Beating_the_speed_of_light_on_the_web/][Google's latest results page]], Dave blew a gasket. [[http://the-edge.blogspot.com/2003/09/paul-vixie-vs-verisign.html][IPv13]] was now a reality!!?? 

So he dusted off the client, and rewrote it (with the help of some friends) to drop the network backend entirely in favor of curl, and incorporated the json API. After some positive feedback the new gnugol client seemed useful enough for general use, so he released it on github.

Whether or not he'll pursue the original gnugold network client remains a question of time and motivation.
* The origin of this name

src/clients/gnugol.c
/* This code is Copyright (C) 2008 by Michael David Taht 
   and released under the terms of the GNU AFFERO PUBLIC_LICENSE, version 3 
   for details of this license see the COPYRIGHT file */

/* Command line client for gnugol */

#define _GNU_SOURCE

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>
#include <getopt.h>

#include "nodelist.h"
#include "query.h"
#include "formats.h"
#include "gnugol_engines.h"
#include "utf8_urlencode.h"

struct  output_types {
  const enum gnugol_formatter id;
  const char *const desc;
};

// FIXME: Verify differences between ikiwiki and media wiki format

List c_engines;

static const struct output_types output_type[] = {
  { FORMATHTML5, "html5" },
  { FORMATHTML, "html" },
  { FORMATIKI, "iki" },
  { FORMATWIKI, "wiki" },
  { FORMATXML,  "xml" },
  { FORMATJSON, "json" },
  { FORMATORG,  "org" },
  { FORMATTERM, "term" },
  { FORMATTEXT, "text" },
  { FORMATSSML, "ssml" },
  { FORMATTEXTILE, "textile" },
  { FORMATRAW,  "raw" },
  { FORMATMDWN, "mdwn" },
  { FORMATMDWN, "md" },
  { FORMATMDWN, "markdown" },
  { FORMATINFO, "info" },
  { FORMATCSV, "csv" },
  { FORMATRAW, "raw" },
  { FORMATSQL, "sql" },
  { FORMATMAN, "man" },
  { FORMATDNS, "dns" },
  { FORMATLISP, "lisp" },
  { 0, NULL },
  };

int usage (char *err) {
  if(err) fprintf(stderr,"%s\n",err);
  printf("gnugol [options] keywords to search for\n");
  printf(
	 "-e --engine        [bing|google|dummy]\n"
	 "-o --output        [html|json|org|mdwn|wiki|text|term|ssml|textile|raw]\n"
	 "-n --nresults      number of results to fetch\n"
	 "-p --position      start of results to fetch\n"
	 "-s --snippets  0|1 disable/enable snippets\n"
	 "-t --titles    0|1 disable/enable titles\n"
	 "-u --urls      0|1 disable/enable urls\n"
	 "-H --header    0|1 disable/enable output header\n"
	 "-F --footer    0|1 disable/enable output footer\n"
	 "-i --indent    X   result indentation level\n"
	 "-S --safe    0|1|2 [off|moderate|active] result filtering\n"
	 "-d --debug     X   debug output level\n"
	 "-U --escaped       input is already url escaped\n"
	 "-l --language-in   [en|es|fr|nl or other 2 char locale]\n"
	 "-L --language-out  [en|es|fr|nl or other 2 char locale]\n"
	 "-v --verbose       provide more verbose insight\n"
	 "-h --help          this message\n"
	 "-a --about         [credits|copyright|license|source|config|manual|stats]\n"

#ifdef HAVE_GNUGOLD
	 "-b --broadcast broadcast results to local network\n"
	 "   --dontfork don't fork off the server\n"
	 "-4 --ipv4 listen on ipv4\n"
	 "-6 --ipv6 listen on ipv6\n"
	 "-m --multicast ask for results from local network\n"
	 "-P --prime     prime the caches, routes, etc\n"
	 "-R --register\n"
	 "-S --Secure    use secure transport\n"
	 "-T --trust networks\n"
#endif

#ifdef WHENIHAVETIMETOADDTHESEOFFICIALLY
	 "-A --ads 0|1\n"
	 "-c --cache     serve only results from cache(s)\n"
	 "-f --force     force a new query, even if cached\n"
	 "-I --input     [filename] input from a file\n"
	 "-5 --offline   store up query for later\n"
	 "-r --reverse   reverse the list\n"
#endif

#ifdef WHATABOUTTHESE
	 "-C --config\n"
	 "-D --defaults\n"
	 "-g --plugin\n"
	 "   --source\n"
#endif
  );

  exit(-1);
}

static const struct option long_options[] = {
  { "about"		, no_argument		, NULL , 'a' } ,
  { "debug"		, required_argument	, NULL , 'd' } ,
  { "engine"		, required_argument	, NULL , 'e' } ,
  { "footer"		, required_argument	, NULL , 'F' } ,
  { "header"		, required_argument	, NULL , 'H' } ,
  { "help"		, no_argument		, NULL , 'h' } ,
  { "indent"		, required_argument	, NULL , 'i' } ,
  { "language-in"	, required_argument	, NULL , 'l' } ,
  { "language-out"	, required_argument	, NULL , 'L' } ,
  { "nresults"		, required_argument	, NULL , 'n' } ,
  { "output"		, required_argument	, NULL , 'o' } ,
  { "position"		, required_argument	, NULL , 'p' } ,
  { "snippets"		, required_argument	, NULL , 's' } ,
  { "safe"		, required_argument    	, NULL , 'S' } ,
  { "titles"		, required_argument	, NULL , 't' } ,
  { "urls"		, required_argument	, NULL , 'u' } ,
  { "escaped"      	, no_argument		, NULL , 'U' } ,
  { "verbose"		, no_argument		, NULL , 'v' } ,

#ifdef HAVE_GNUGOLD
  { "broadcast"		, no_argument		, NULL , 'b' } ,
  { "dontfork"		, no_argument		, NULL ,  0  } ,
  { "ipv4"		, no_argument		, NULL , '4' } ,
  { "ipv6"		, no_argument		, NULL , '6' } ,
  { "multicast"		, no_argument		, NULL , 'm' } ,
  { "prime"		, no_argument		, NULL , 'P' } ,
  { "register"		, no_argument		, NULL , 'R' } ,
  { "secure"		, no_argument		, NULL , 'S' } ,
  { "trust"		, no_argument		, NULL , 'T' } ,
#endif

#ifdef WHENIHAVETIMETOADDTHESEOFFICIALLY
  { "ads"		, required_argument	, NULL , 'A' } ,
  { "cache"		, no_argument		, NULL , 'c' } ,
  { "force"		, no_argument		, NULL , 'f' } ,
  { "input"		, required_argument	, NULL , 'I' } ,
  { "offline"		, no_argument		, NULL , '5' } ,
  { "reverse"		, no_argument		, NULL , 'r' } ,
#endif

#ifdef WHATABOUTTHESE
  { "word-break"	, no_argument		, NULL , 'B' } ,
  { "config"		, no_argument		, NULL , 'C' } ,
  { "defaults"		, no_argument		, NULL , 'D' } ,
  { "plugin"		, no_argument		, NULL , 'g' } ,
  { "source"		, no_argument		, NULL ,  0  } ,
#endif

  { NULL		, 0			, NULL ,  0  }
};

int gnugol_parse_config_file(QueryOptions_t *q __attribute__((unused))) {
  return 0;
}

#define penabled(a) if(o->a) fprintf(fp,"" # a " ");

int
print_enabled_options(QueryOptions_t *o, FILE *fp) {
  if(o->verbose) fprintf(fp,"Search Keywords: %s\n",o->keywords);
  fprintf(fp,"Results Requested: %d\n", o->nresults);
  fprintf(fp,"Starting position: %d\n",o->position);
  fprintf(fp,"Enabled Options: ");
  penabled(header);
  penabled(footer);
  penabled(urls);
  penabled(titles);
  penabled(snippets);
  penabled(ads);
  penabled(misc);
  penabled(reverse);
#ifdef HAVE_GNUGOLD
  penabled(prime);
  penabled(broadcast);
  penabled(multicast);
  penabled(ipv4);
  penabled(ipv6);
#endif
  penabled(force);
  penabled(cache);
  penabled(offline);
  penabled(safe);
  penabled(reg);
  penabled(indent);
  penabled(engine);
  penabled(mirror);
  penabled(plugin);
  penabled(debug);
  fprintf(fp,"\n");
  return 0;
}

#define BOOLOPT(OPTION) OPTION = (strtoul(optarg,NULL,10) & 1)
#ifdef HAVE_GNUGOLD
#  define QSTRING "ad:e:F:H:hi:l:L:n:o:p:s:S:t:u:Uvb46mPRST"
#else
#  define QSTRING "ad:e:F:H:hi:l:L:n:o:p:s:S:t:u:Uv"
#endif

int process_options(int argc, char **argv, QueryOptions_t *o) 
{
  int          option_index;
  int          opt;
  GnuGolEngine engine;

  option_index = 0;
  opt          = 0;
  optind       = 1;

  while(true)
  {
    opt = getopt_long(argc, argv,
		      QSTRING,
		      long_options, &option_index);
    if(opt == -1) break;

    switch (opt)
    {
      case 'a': o->about = 1;  break;
      case 'd': o->debug = strtoul(optarg,NULL,10); break;
      case 'e':
           o->engine = 1;
           engine = gnugol_engine_load(optarg);
           if (engine == NULL)
             fprintf(stderr,"engine %s not supported\n",optarg);
           else
             ListAddTail(&c_engines,&engine->node);
           break;
      case 'F': BOOLOPT(o->footer); break;
      case 'H': o->header = strtoul(optarg,NULL,10); break;
      case 'h':
      case '?': usage(NULL); break;
      case 'i': o->indent = strtoul(optarg,NULL,10); break;
      case 'l': strcpy(o->input_language,optarg); break;
      case 'L': strcpy(o->output_language,optarg); break;
      case 'n': o->nresults = strtoul(optarg,NULL,10); break;
      case 'o':
           for(int i = 0; output_type[i].desc != NULL; i++)
	    if(strcmp(output_type[i].desc,optarg) == 0)
	      o->format = output_type[i].id;
           break;
      case 'p': o->position = strtoul(optarg,NULL,10); break;
      case 's': o->snippets = strtoul(optarg,NULL,10); break;
      case 'S': o->safe = strtoul(optarg,NULL,10); break;
      case 't': o->titles = strtoul(optarg,NULL,10); break;
      case 'u': o->urls = strtoul(optarg,NULL,10); break;
      case 'U': o->url_escape = 1; break;
      case 'v': o->verbose = 1; break;

#ifdef HAVE_GNUGOLD
      case 'b': o->broadcast = 1; break;
      /* dontfork missing */
      case '4': o->ipv4 = 1; break;
      case '6': o->ipv6 = 1; break;
      case 'm': o->multicast = 1; break;
      /* prime missing */
      case 'R': o->reg = 1; break;
      case 'Z': o->secure = 1; break; // unimplemented
      case 'T': o->trust = 1; break;
#endif

#ifdef WHENIHAVETIMETOADDTHESEOFFICIALLY
      case 'A': o->ads = strtoul(optarg,NULL,10); break;
      case 'c': o->cache = 1; break;
      /* force missing */
      case 'I': o->input = 1; o->input_file = optarg; break; // FIXME
      case '5': o->offline = 1; break;
      case 'r': o->reverse = 1; break;
#endif

#ifdef WHATABOUTTHESE
      /* config missing */
      /* defaults missing */
      case 'P': o->plugin = 1; break;
      case 'f': o->output = 1; break; // FIXME
      /* source missing */
#endif

      default: fprintf(stderr,"%d",opt); usage("Invalid option"); break;
    }
  } 

  return optind;
}

void finish_setup(QueryOptions_t *o,int idx,int argc,char **argv)
{
  GnuGolEngine engine;
  char    string[MAX_MTU];
  size_t  querylen = 0;
  int     i;

  string[0] = '\0';
  
  if (!o->about && ListEmpty(&c_engines))  
  {
    engine = gnugol_engine_load(o->engine_name);
    if (engine == NULL)
    {
      fprintf(stderr,"default engine not found!  Panic!\n");
      exit(EXIT_FAILURE);
    }
    
    ListAddTail(&c_engines,&engine->node);
  }
  
  if (o->about)
  {
    engine = gnugol_engine_load("credits");
    if (engine == NULL)
    {
      fprintf(stderr,"default engine not found!  Panic!\n");
      exit(EXIT_FAILURE);
    }
    
    ListAddTail(&c_engines,&engine->node);
  }

  for(i = idx; i < argc; i++) {
	  if((querylen += (strlen(argv[i])+1) > MAX_MTU - 80)) {
		  fprintf(stderr,"Too many words in query, try something smaller\n");
		  exit(EXIT_FAILURE);
	  }
	  /* FIXME: Although I did a length check above it could be cleaner here */
	  if(!o->url_escape) {
		  strcat(string,argv[i]);
		  if(i+1 < argc) strcat(string," ");
	  } else {
		  strcat(string,argv[i]);
		  if(i+1 < argc) strcat(string,"+");
	  }
  }

  if(!o->url_escape) {
	  url_escape_utf8(o->keywords,string);
	  o->url_escape = 1;
  } else {
	  strcpy(o->keywords,string);
  }
  if(o->debug > 0) print_enabled_options(o, stderr);
  if(!(o->urls | o->snippets | o->ads | o->titles)) o->urls = 1;
  if(o->about) {
    o->engine_name = "credits";
    o->header_str = "About: ";
  }
}

static void gnugol_default_language (QueryOptions_t *q) {
	// Get this from the locale? What's portable?
	// FIXME: We always want utf-8

	char *lang = getenv("LANG");
	if(lang != NULL) {
		strcpy(q->input_language,lang);
	} else {
		strcpy(q->input_language,"en_US.utf8");
	}
}

static void gnugol_default_QueryOptions(QueryOptions_t *q) {
	q->nresults = 5;
	q->position = 0;
	q->urls = 1;
	q->snippets = 1;
	q->titles = 1;
	q->engine_name = "google";
	q->header = 1;
	q->footer = 1;
	q->format = FORMATDEFAULT; // ORG
	q->indent = -1; // None
	q->safe = 1; // Moderate
	gnugol_default_language(q);
}

void process_environ(char *program,QueryOptions_t *query)
{
  char   *opt;
  size_t  optlen;
  
  assert(program != NULL);
  assert(query   != NULL);
  
  opt = getenv("GNUGOL_OPTS");
  if (opt == NULL) return;
  optlen = strlen(opt) + 1;
  
  char    optcopy[optlen];
  char   *argv[optlen + 1];
  size_t  argc;
  char   *p;
  
  memcpy(optcopy,opt,optlen);
  p       = optcopy;
  argv[0] = program;
  
  for (argc = 1 ; ; )
  {
    argv[argc] = strtok(p," \t\v\r\n");
    if (argv[argc] == NULL) break;
    argc++;
    p = NULL;
  }
  
  process_options(argc,argv,query);
}

int main(int argc, char **argv) {
  int            result;
  int            words;
  QueryOptions_t master;
  QueryOptions_t q;
  GnuGolEngine   engine;

  ListInit(&c_engines);

  gnugol_init_QueryOptions(&master);
  gnugol_default_QueryOptions(&master);
  
  process_environ(argv[0],&master);
  words = process_options(argc,argv,&master);
  finish_setup(&master,words,argc,argv);
  
  assert(!ListEmpty(&c_engines));
  
  for(
        engine = (GnuGolEngine)ListGetHead(&c_engines);
        NodeValid(&engine->node);
        engine = (GnuGolEngine)NodeNext(&engine->node)
      )
  {
    q      = master;
    result = gnugol_engine_query(engine,&q);

    if(q.returned_results > 0) {
      printf("%s",q.out.s);
    }

    if(result < 0 || q.debug > 5) {
      fprintf(stderr,"Errors: %s\nWarnings:%s\n",q.err.s,q.wrn.s);
    }

    if(q.debug > 10) {
      fprintf(stderr,"out len = %d\n size = %d, Contents = %s\n",q.out.len, q.out.size, q.out.s);
      fprintf(stderr,"wrn len = %d\n size = %d, Contents = %s\n",q.wrn.len, q.wrn.size, q.wrn.s);
      fprintf(stderr,"err len = %d\n size = %d, Contents  = %s\n",q.err.len, q.err.size, q.err.s);
    }
    /*gnugol_free_QueryOptions(&q);*/
  }
  
  for(
       engine = (GnuGolEngine)ListRemTail(&c_engines);
       NodeValid(&engine->node);
       engine = (GnuGolEngine)ListRemTail(&c_engines)
     )
  {
    gnugol_engine_unload(engine);
  }

  return(0);
}
